### ISSUE TITLE
How to add audio of viewer?

### ISSUE THREAD

**Issue by libdup** [muaz-khan/Chrome-Extensions#62]
I want to add audio of the viewer so that both the parties( sharer and viewer) can talk to each other. In which file do I have to make changes. I am new to webRTC so any help will be appreciated. Thanks in advance.

**Comment by muaz-khan** [comment]
**First step**, open [this file](https://github.com/muaz-khan/Chrome-Extensions/blob/master/desktopCapture-p2p/index.html) and search for `connection.session`. Remove `{oneway:true}`. Now `connection.session` looks like this:

```javascript
// do not set {video:true} because you are trying to share viewer's mic only
connection.session = {
   audio: true
};
```

**Second step**, open [this file](https://github.com/muaz-khan/Chrome-Extensions/blob/master/desktopCapture-p2p/background/shareStreamUsingRTCMultiConnection.js) and search for same `connection.session` object.  Simply remove `oneway:true`.

```javascript
connection.session = {
   audio: true,
   video: true
};
```

**Third step**, find `OfferToReceiveAudio` in the second step's file. Set `true` bot for `OfferToReceiveAudio` and `OfferToReceiveVideo`. Now It looks like this:

```javascript
connection.sdpConstraints.mandatory = {
    OfferToReceiveAudio: true,
    OfferToReceiveVideo: true
};
```

**Fourth step**, in the same second step's file, search for `connection.onstream` and remove that function declaration. Currently it is trying to ignore/pause/stop all incoming streams but we need to receive mic from the viewer. So we need to remove `connection.onstream` declaration or otherwise make it look same like this:
* https://www.rtcmulticonnection.org/docs/onstream/

----

### To recap:

1. First of all, you need to modify viewer's `index.html` file and remove `oneway:true, video: true`.
2. Now remove `{oneway:true}` on extension's codes as well.
3. By default extension tries to ignore all kinds of incoming streams. To bypass this behavior, you need to set `OfferToReceiveAudio:true` and `OfferToReceiveVideo:true` .
4. You also need to either remove `onstream` event or make it look like this:

```javascript
connection.onstream = function(event) {
    var video = event.mediaElement;
    document.body.appendChild(video);
};
```

If above code throws error then you can try this instead:

```javascript
connection.onstream = function(event) {
    var video = event.mediaElement;
    video.play();
};
```

**Comment by muaz-khan** [comment]
Please add following into the index.html:

```javascript
connection.mediaConstraints = {
    audio: true
};
```

Please make sure that you're using HTTPs. Because one can not access mic on HTTP until if [this plugin](https://github.com/muaz-khan/Chrome-Extensions/tree/master/getUserMedia-on-http) is installed.

`mediaConstrains` documentation: https://www.rtcmulticonnection.org/docs/mediaConstraints/

Still facing issues? Please make sure that microphone isn't blocked by Chrome:

```sh
chrome://settings/content/microphone?search=microphone
```

You can remove all camera-blocking as well:

```sh
chrome://settings/content/camera?search=camera
```

**Comment by libdup** [comment]
@muaz-khan , I created ssl certificates using openSSL to use https on localhost. It is working fine on my system. But as i try to access the url from another system on local LAN, the connection keeps reloading quickly and video doesn't appear at all.
  If i access the url simultaneously from my system and the other, as the other system joins reloading starts happen on my system too. What could be the problem?

**Comment by libdup** [comment]
This error comes when reloading is happen from another system
![error on reloading](https://user-images.githubusercontent.com/46367863/51013527-f3b56e00-1588-11e9-9d16-e36c216b85d9.png)

**Comment by vvug** [comment]
@muaz-khan do you think it is possible to add this as an option to the extension? I think it is very useful, as a common use use case is to record a videoconference for future reference.

**Comment by vvug** [comment]
Or, as another example, record a video (a tab) while also watching and listening to it.

**Comment by muaz-khan** [comment]
@vvug I've already added the CHAT option.  I'll add microphone/camera sharing in the next commit. 

![desktopcapture-p2p](https://user-images.githubusercontent.com/477092/51313913-874adb00-1a70-11e9-88e9-09a190bcd0f4.png)
